sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -557,6 +581,31 @@ case class Sin(child: Expression) extends UnaryMathExpression(math.sin, "SIN")
   """)
 case class Sinh(child: Expression) extends UnaryMathExpression(math.sinh, "SINH")
 
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr) - Returns inverse hyperbolic sine of `expr`.
+  """,
+  arguments = """
+    Arguments:
+      * expr - hyperbolic angle
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(0);
+       0.0
+  """,
+  since = "3.0.0")
+case class Asinh(child: Expression)
+  extends UnaryMathExpression((x: Double) => x match {
+    case Double.NegativeInfinity => Double.NegativeInfinity
+    case _ => math.log(x + math.sqrt(x * x + 1.0)) }, "ASINH") {
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    defineCodeGen(ctx, ev, c =>
+      s"${ev.value} = $c == Double.NEGATIVE_INFINITY ? Double.NEGATIVE_INFINITY : " +
```

Comment:
Hmm. This still might fail in some case. `($c)` will be better?

Reply:
what about rewriting this as an `if`? I think it is safer indeed. I had some issues with this syntax in another PR.

Reply:
Ya. I'm worrying the same situation. $c is not safe in general.

Reply:
Moreover, when rewriting this as an `if`, you can't use `defineCodeGen` anymore